pkg/objstore/oci/helper.go

```diff
@@ -0,0 +1,327 @@
+		}
+
+		if objstore.ApplyIterOptions(options...).Recursive {
+			for _, objectName := range objectNames {
+				if strings.HasSuffix(objectName, DirDelim) {
+					subObjectNames, err := listAllObjects(ctx, bkt, objectName, options...)
+					if err != nil {
+						return nil, err
+					}
+					allObjectNames = append(allObjectNames, subObjectNames...)
+				} else {
+					allObjectNames = append(allObjectNames, objectName)
+				}
+			}
+		} else {
+			allObjectNames = append(allObjectNames, objectNames...)
+		}
+	}
+	return allObjectNames, nil
+}
+
+func listObjects(ctx context.Context, bkt Bucket, prefix string, start *string) (objectNames []string, nextStartWith *string, err error) {
+	request := objectstorage.ListObjectsRequest{
+		NamespaceName:   &bkt.namespace,
+		BucketName:      &bkt.name,
+		Delimiter:       common.String(DirDelim),
+		Prefix:          &prefix,
+		Start:           start,
+		RequestMetadata: bkt.requestMetadata,
+	}
+	response, err := bkt.client.ListObjects(ctx, request)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, object := range response.ListObjects.Objects {
+		objectNames = append(objectNames, *object.Name)
+	}
+	objectNames = append(objectNames, response.ListObjects.Prefixes...)
+
+	return objectNames, response.NextStartWith, nil
+}
+
+func uploadObject(ctx context.Context, bkt Bucket, objectName string, reader io.Reader) (err error) {
+	req := transfer.UploadStreamRequest{
+		UploadRequest: transfer.UploadRequest{
+			NamespaceName:                       common.String(bkt.namespace),
+			BucketName:                          common.String(bkt.name),
+			ObjectName:                          &objectName,
+			EnableMultipartChecksumVerification: common.Bool(true), // TODO: should we check?
+			ObjectStorageClient:                 bkt.client,
+			RequestMetadata:                     bkt.requestMetadata,
+		},
+		StreamReader: reader,
+	}
+	if bkt.partSize > 0 {
+		req.UploadRequest.PartSize = &bkt.partSize
+	}
+
+	uploadManager := transfer.NewUploadManager()
+	_, err = uploadManager.UploadStream(ctx, req)
+	return
+}
+
+func deleteObject(ctx context.Context, bkt Bucket, objectName string) (err error) {
+	request := objectstorage.DeleteObjectRequest{
+		NamespaceName:   &bkt.namespace,
+		BucketName:      &bkt.name,
+		ObjectName:      &objectName,
+		RequestMetadata: bkt.requestMetadata,
+	}
+	_, err = bkt.client.DeleteObject(ctx, request)
+	return
+}
+
+func createBucket(ctx context.Context, bkt *Bucket, compartmentId string) (err error) {
+	request := objectstorage.CreateBucketRequest{
+		NamespaceName:   &bkt.namespace,
+		RequestMetadata: bkt.requestMetadata,
+	}
+	request.CompartmentId = &compartmentId
+	request.Name = &bkt.name
+	request.Metadata = make(map[string]string)
+	request.PublicAccessType = objectstorage.CreateBucketDetailsPublicAccessTypeNopublicaccess
+	_, err = bkt.client.CreateBucket(ctx, request)
+	return
+}
+
+func deleteBucket(ctx context.Context, bkt *Bucket) (err error) {
```

Comment:
In fact, why do we separate functions for this? Just move this into `Bucket.deleteBucket()` and that's all (:

Reply:
Same reply as above